File: src/fast_resume/adapters/kiro.py
"""Kiro CLI session adapter.

Kiro stores each session as a pair of files in `~/.kiro/sessions/cli/`:

    <uuid>.json    Metadata: session_id, cwd, created_at, updated_at, title.
    <uuid>.jsonl   Event stream with kinds: Prompt, AssistantMessage, ToolResults.

The metadata file already carries enough to render a row; the JSONL is parsed
for full message content so search can match conversation text.
"""

import orjson
from datetime import datetime
from pathlib import Path

from ..config import AGENTS, KIRO_DIR
from ..logging_config import log_parse_error
from .base import BaseSessionAdapter, ErrorCallback, ParseError, Session, truncate_title


class KiroAdapter(BaseSessionAdapter):
    """Adapter for Kiro CLI sessions."""

    name = "kiro"
    color = AGENTS["kiro"]["color"]
    badge = AGENTS["kiro"]["badge"]
    supports_yolo = True

    def __init__(self, sessions_dir: Path | None = None) -> None:
        self._sessions_dir = sessions_dir if sessions_dir is not None else KIRO_DIR

    def find_sessions(self) -> list[Session]:
        """Find all Kiro CLI sessions."""
        if not self.is_available():
            return []

        sessions = []
        for meta_file in self._sessions_dir.glob("*.json"):
            session = self._parse_session_file(meta_file)
            if session:
                sessions.append(session)
        return sessions

    def _parse_session_file(
        self, session_file: Path, on_error: ErrorCallback = None
    ) -> Session | None:
        """Parse a Kiro session given its `<uuid>.json` metadata path.

        The matching `<uuid>.jsonl` (if present) is read for message content.
        """
        meta_file = session_file
        try:
            with open(meta_file, "rb") as f:
                meta = orjson.loads(f.read())
        except orjson.JSONDecodeError as e:
            self._report(on_error, str(meta_file), "JSONDecodeError", str(e))
            return None
        except OSError as e:
            self._report(on_error, str(meta_file), "OSError", str(e))
            return None

        try:
            session_id = meta.get("session_id") or meta_file.stem
            directory = meta.get("cwd", "")
            title_from_meta = (meta.get("title") or "").strip()

            timestamp = self._parse_timestamp(
                meta.get("updated_at") or meta.get("created_at"), meta_file
            )

            events_file = meta_file.with_suffix(".jsonl")
            messages, first_user_prompt, turn_count = self._parse_events(
                events_file, on_error=on_error
            )

            title = title_from_meta or first_user_prompt
            if title:
                title = truncate_title(title, max_length=80, word_break=False)
            else:
                title = "Kiro session"

            full_content = "\n\n".join(messages)

            return Session(
                id=session_id,
                agent=self.name,
                title=title,
                directory=directory,
                timestamp=timestamp,
                content=full_content,
                message_count=turn_count,
            )
        except (KeyError, TypeError, AttributeError) as e:
            self._report(on_error, str(meta_file), type(e).__name__, str(e))
            return None

    def _parse_events(
        self, events_file: Path, on_error: ErrorCallback = None
    ) -> tuple[list[str], str, int]:
        """Parse the `.jsonl` event stream alongside the metadata file.

        Returns:
            (messages, first_user_prompt, turn_count)
        """
        messages: list[str] = []
        first_user_prompt = ""
        turn_count = 0

        if not events_file.exists():
            return messages, first_user_prompt, turn_count

        try:
            with open(events_file, "rb") as f:
                for line in f:
                    line = line.strip()
                    if not line:
                        continue
                    try:
                        entry = orjson.loads(line)
                    except orjson.JSONDecodeError:
                        continue

                    kind = entry.get("kind", "")
                    data = entry.get("data", {})
                    if kind == "Prompt":
                        text = self._extract_text(data.get("content", []))
                        if text:
                            messages.append(f"» {text}")
                            turn_count += 1
                            if not first_user_prompt:
                                first_user_prompt = text
                    elif kind == "AssistantMessage":
                        text = self._extract_text(data.get("content", []))
                        if text:
                            messages.append(f"  {text}")
                            turn_count += 1
                    # ToolResults are intentionally excluded from the index.
        except OSError as e:
            # Permission-denied / IO failure on the event stream produces an
            # empty session content; surface it so the caller knows parsing
            # was partial rather than the session actually being empty.
            self._report(on_error, str(events_file), "OSError", str(e))

        return messages, first_user_prompt, turn_count

    @staticmethod
    def _extract_text(content: list) -> str:
        """Collect plain-text segments from a Kiro content list, ignoring tool parts."""
        parts: list[str] = []
        for item in content:
            if not isinstance(item, dict):
                continue
            if item.get("kind") != "text":
                continue
            text = item.get("data", "")
            if isinstance(text, str) and text.strip():
                parts.append(text)
        return "\n".join(parts).strip()

    @staticmethod
    def _parse_timestamp(value: str | None, fallback_file: Path) -> datetime:
        if value:
            try:
                # Kiro emits RFC3339 with trailing 'Z'.
                ts = value.replace("Z", "+00:00") if value.endswith("Z") else value
                dt = datetime.fromisoformat(ts)
                if dt.tzinfo is not None:
                    dt = dt.astimezone().replace(tzinfo=None)
                return dt
            except ValueError:
                pass
        return datetime.fromtimestamp(fallback_file.stat().st_mtime)

    def _report(
        self,
        on_error: ErrorCallback,
        file_path: str,
        error_type: str,
        message: str,
    ) -> None:
        error = ParseError(
            agent=self.name,
            file_path=file_path,
            error_type=error_type,
            message=message,
        )
        log_parse_error(error.agent, error.file_path, error.error_type, error.message)
        if on_error:
            on_error(error)

    def _scan_session_files(self) -> dict[str, tuple[Path, float]]:
        """Scan Kiro session metadata files.

        Uses the newer of the meta + events mtimes so that index updates fire
        when only the JSONL grows (a still-running session).
        """
        current: dict[str, tuple[Path, float]] = {}
        for meta_file in self._sessions_dir.glob("*.json"):
            try:
                mtime = meta_file.stat().st_mtime
            except OSError:
                continue
            events_file = meta_file.with_suffix(".jsonl")
            try:
                events_mtime = events_file.stat().st_mtime
                if events_mtime > mtime:
                    mtime = events_mtime
            except OSError:
                pass
            current[meta_file.stem] = (meta_file, mtime)
        return current

    def get_resume_command(self, session: Session, yolo: bool = False) -> list[str]:
        """Get command to resume a Kiro CLI session."""
        cmd = ["kiro-cli", "chat"]
        if yolo:
            cmd.append("--trust-all-tools")
        cmd.extend(["--resume-id", session.id])
        return cmd

File: src/adapters/kiro.rs
/// Kiro CLI session adapter.
///
/// Faithfully ported from python/fast_resume/adapters/kiro.py.
/// Sessions live under ~/.kiro/sessions/cli/ as <uuid>.json + <uuid>.jsonl pairs.
use std::collections::HashMap;
use std::io::{BufRead, BufReader};
use std::path::{Path, PathBuf};
use std::time::UNIX_EPOCH;

use serde_json::Value;

use crate::adapters::{file_based_incremental, AgentAdapter, ErrorCb, IncrementalResult, SessionCb};
use crate::config;
use crate::session::{ParseError, RawAdapterStats, Session};

fn system_time_to_f64(t: std::time::SystemTime) -> f64 {
    t.duration_since(UNIX_EPOCH)
        .map(|d| d.as_secs_f64())
        .unwrap_or(0.0)
}

fn f64_to_timestamp(secs: f64) -> jiff::Timestamp {
    let whole = secs as i64;
    let nanos = ((secs - whole as f64) * 1_000_000_000.0) as i32;
    jiff::Timestamp::new(whole, nanos).unwrap_or(jiff::Timestamp::UNIX_EPOCH)
}

/// Parse an ISO-8601 / RFC-3339 string to a jiff Timestamp.
fn parse_iso_timestamp(s: &str, fallback_mtime: f64) -> jiff::Timestamp {
    let s = if let Some(without_z) = s.strip_suffix('Z') {
        format!("{}+00:00", without_z)
    } else {
        s.to_string()
    };
    if let Ok(ts) = s.parse::<jiff::Timestamp>() {
        return ts;
    }
    f64_to_timestamp(fallback_mtime)
}

fn truncate_title(text: &str, max_length: usize) -> String {
    let text = text.trim();
    if text.len() <= max_length {
        return text.to_string();
    }
    format!("{}...", &text[..max_length])
}

/// Collect plain-text segments from a Kiro content array.
fn extract_text(content: &Value) -> String {
    let mut parts: Vec<&str> = Vec::new();
    if let Value::Array(items) = content {
        for item in items {
            if let Value::Object(obj) = item {
                if obj.get("kind").and_then(|v| v.as_str()) != Some("text") {
                    continue;
                }
                if let Some(text) = obj.get("data").and_then(|v| v.as_str()) {
                    let text = text.trim();
                    if !text.is_empty() {
                        parts.push(text);
                    }
                }
            }
        }
    }
    parts.join("\n")
}

pub struct KiroAdapter {
    sessions_dir: PathBuf,
}

impl KiroAdapter {
    pub fn new() -> Self {
        Self {
            sessions_dir: config::kiro_dir(),
        }
    }

    pub fn with_dir(sessions_dir: PathBuf) -> Self {
        Self { sessions_dir }
    }

    fn scan(&self) -> Vec<(String, PathBuf, f64)> {
        let mut results = Vec::new();

        let read_dir = match std::fs::read_dir(&self.sessions_dir) {
            Ok(rd) => rd,
            Err(_) => return results,
        };

        for entry in read_dir.flatten() {
            let path = entry.path();
            if path.extension().and_then(|e| e.to_str()) != Some("json") {
                continue;
            }
            if !path.is_file() {
                continue;
            }

            let meta_mtime = match path.metadata().and_then(|m| m.modified()) {
                Ok(t) => system_time_to_f64(t),
                Err(_) => continue,
            };

            // Use the later of .json or .jsonl mtime.
            let events_file = path.with_extension("jsonl");
            let mtime = if events_file.exists() {
                let ev_mtime = events_file
                    .metadata()
                    .and_then(|m| m.modified())
                    .map(system_time_to_f64)
                    .unwrap_or(0.0);
                meta_mtime.max(ev_mtime)
            } else {
                meta_mtime
            };

            let stem = path
                .file_stem()
                .and_then(|s| s.to_str())
                .unwrap_or("")
                .to_string();

            results.push((stem, path, mtime));
        }

        results
    }

    pub fn parse_session_file(
        &self,
        path: &Path,
        on_error: Option<ErrorCb<'_>>,
    ) -> Option<Session> {
        match self.parse_meta_file(path) {
            Ok(session) => session,
            Err(e) => {
                if let Some(cb) = on_error {
                    cb(ParseError {
                        agent: self.name().to_string(),
                        file_path: path.display().to_string(),
                        error_type: "IOError".to_string(),
                        message: e.to_string(),
                    });
                }
                None
            }
        }
    }

    fn parse_meta_file(&self, meta_file: &Path) -> anyhow::Result<Option<Session>> {
        let meta_mtime = meta_file
            .metadata()
            .and_then(|m| m.modified())
            .map(system_time_to_f64)
            .unwrap_or(0.0);

        // meta_file comes from adapter's own scan; not user-supplied.
        // nosemgrep: rust.actix.path-traversal.tainted-path.tainted-path
        let meta_content = std::fs::read_to_string(meta_file)?;
        let meta: Value = serde_json::from_str(&meta_content)?;

        let session_id = meta
            .get("session_id")
            .and_then(|v| v.as_str())
            .unwrap_or_else(|| {
                meta_file
                    .file_stem()
                    .and_then(|s| s.to_str())
                    .unwrap_or("")
            })
            .to_string();

        let directory = meta.get("cwd").and_then(|v| v.as_str()).unwrap_or("").to_string();

        let title_from_meta = meta
            .get("title")
            .and_then(|v| v.as_str())
            .unwrap_or("")
            .trim()
            .to_string();

        let ts_str = meta
            .get("updated_at")
            .or_else(|| meta.get("created_at"))
            .and_then(|v| v.as_str())
            .unwrap_or("");

        let timestamp = if !ts_str.is_empty() {
            parse_iso_timestamp(ts_str, meta_mtime)
        } else {
            f64_to_timestamp(meta_mtime)
        };

        // Parse the .jsonl events file.
        let events_file = meta_file.with_extension("jsonl");
        let mut messages: Vec<String> = Vec::new();
        let mut first_user_prompt = String::new();
        let mut turn_count: u32 = 0;

        if events_file.exists() {
            // nosemgrep: rust.actix.path-traversal.tainted-path.tainted-path
            let file = std::fs::File::open(&events_file)?;
            let reader = BufReader::new(file);

            for line in reader.lines() {
                let line = match line {
                    Ok(l) => l,
                    Err(_) => continue,
                };
                let trimmed = line.trim();
                if trimmed.is_empty() {
                    continue;
                }
                let entry: Value = match serde_json::from_str(trimmed) {
                    Ok(v) => v,
                    Err(_) => continue,
                };

                let kind = entry.get("kind").and_then(|v| v.as_str()).unwrap_or("");
                let data = entry.get("data").cloned().unwrap_or(Value::Null);

                if kind == "Prompt" {
                    let text = extract_text(data.get("content").unwrap_or(&Value::Array(vec![])));
                    if !text.is_empty() {
                        messages.push(format!("» {}", text));
                        turn_count += 1;
                        if first_user_prompt.is_empty() {
                            first_user_prompt = text;
                        }
                    }
                } else if kind == "AssistantMessage" {
                    let text = extract_text(data.get("content").unwrap_or(&Value::Array(vec![])));
                    if !text.is_empty() {
                        messages.push(format!("  {}", text));
                        turn_count += 1;
                    }
                }
                // ToolResults are intentionally excluded.
            }
        }

        let raw_title = if !title_from_meta.is_empty() {
            title_from_meta
        } else if !first_user_prompt.is_empty() {
            first_user_prompt
        } else {
            "Kiro session".to_string()
        };
        let title = truncate_title(&raw_title, 80);

        let events_mtime = events_file
            .metadata()
            .and_then(|m| m.modified())
            .map(system_time_to_f64)
            .unwrap_or(0.0);
        let mtime = meta_mtime.max(events_mtime);

        Ok(Some(Session {
            id: session_id,
            agent: self.name().to_string(),
            title,
            directory,
            timestamp,
            content: messages.join("\n\n"),
            message_count: turn_count,
            mtime,
            yolo: false,
        }))
    }
}

impl Default for KiroAdapter {
    fn default() -> Self {
        Self::new()
    }
}

impl AgentAdapter for KiroAdapter {
    fn name(&self) -> &'static str {
        "kiro"
    }

    fn color(&self) -> &'static str {
        "#5C1FFB"
    }

    fn badge(&self) -> &'static str {
        "kiro"
    }

    fn supports_yolo(&self) -> bool {
        true
    }

    fn is_available(&self) -> bool {
        self.sessions_dir.exists()
    }

    fn find_sessions(&self) -> Vec<Session> {
        if !self.is_available() {
            return Vec::new();
        }
        self.scan()
            .into_iter()
            .filter_map(|(_, path, _)| self.parse_session_file(&path, None))
            .collect()
    }

    fn find_sessions_incremental(
        &self,
        known: &HashMap<String, (f64, String)>,
        on_error: Option<ErrorCb<'_>>,
        on_session: Option<SessionCb<'_>>,
    ) -> IncrementalResult {
        file_based_incremental(
            self.name(),
            self.is_available(),
            known,
            || self.scan(),
            |path, err_cb| self.parse_session_file(path, err_cb),
            on_error,
            on_session,
        )
    }

    fn get_resume_command(&self, session: &Session, yolo: bool) -> Vec<String> {
        let mut cmd = vec!["kiro-cli".to_string(), "chat".to_string()];
        if yolo {
            cmd.push("--trust-all-tools".to_string());
        }
        cmd.push("--resume-id".to_string());
        cmd.push(session.id.clone());
        cmd
    }

    fn get_raw_stats(&self) -> RawAdapterStats {
        if !self.is_available() {
            return RawAdapterStats {
                agent: self.name().to_string(),
                data_dir: self.sessions_dir.display().to_string(),
                available: false,
                file_count: 0,
                total_bytes: 0,
            };
        }
        let files = self.scan();
        let file_count = files.len() as u64;
        let total_bytes: u64 = files
            .iter()
            .map(|(_, meta_path, _)| {
                let meta_bytes = meta_path.metadata().map(|m| m.len()).unwrap_or(0);
                let ev_bytes = meta_path
                    .with_extension("jsonl")
                    .metadata()
                    .map(|m| m.len())
                    .unwrap_or(0);
                meta_bytes + ev_bytes
            })
            .sum();
        RawAdapterStats {
            agent: self.name().to_string(),
            data_dir: self.sessions_dir.display().to_string(),
            available: true,
            file_count,
            total_bytes,
        }
    }
}
